### DISCUSSION TITLE
Getting the text of a fenced Code Block

### DISCUSSION THREAD

**Discussion by saminahbab** [yuin/goldmark#398] [Q&A]
It really is not clear to me how to retrieve the text of a fenced code block. I have the following code: 

```
func Parse(source string) error {
	md := goldmark.New()
	document := md.Parser().Parse(text.NewReader([]byte(source)))
	document.Dump([]byte(source), 4)
	parseChildren(document, source)
	return nil
}

// parse the tree looking specifically for text and code blocks
// and file names
func parseChildren(node ast.Node, source string) {
	pointer := node.FirstChild()

	for pointer != nil {
		println(pointer.Kind().String())
		text := pointer.Text([]byte(source))

		fmt.Printf("%s\n", text)
		pointer = pointer.NextSibling()
	}

}

```

But the pointer only shows the Text for Paragraph blocks, when there are CodeBlocks, it returns nothing. 

Is there anyone who has an idea of how to retrieve text of code blocks? 
When I run document.Dump(source), it is clear that it can parse the text and display it. 
@yuin

**Comment by saminahbab** [comment]
So far I have this (very messy)

```
func parseChildren(node ast.Node, source string) {
	pointer := node.FirstChild()
	sourceBytes := []byte(source)
	for pointer != nil {
		println(pointer.Kind().String())
		text := pointer.Text([]byte(source))

		fmt.Printf("%s\n", text)
		println(pointer.Lines().Len())
		for i := 0; i < pointer.Lines().Len(); i++ {
			start := pointer.Lines().At(i).Start
			stop := pointer.Lines().At(i).Stop
			fmt.Printf("%s\n", sourceBytes[start:stop])
		}
		pointer = pointer.NextSibling()
	}

```

which is showing each line of the code block, but surely there should be some kind of convenience method that allows for this?

**Comment by abhinav** [comment]
You need to walk the AST to search for the `*ast.FencedCodeBlock`.

```go
ast.Walk(document, func(node ast.Node, entering bool) (ast.WalkStatus, error) {
	if !entering {
		return ast.WalkContinue, nil
	}

	cb, ok := node.(*ast.FencedCodeBlock)
	if !ok {
		return ast.WalkContinue, nil
	}

	// do stuff with FencedCodeBlock

	return ast.WalkContinue, nil
})
```

Once you've found it, use `Node.Lines()` to get the `*text.Segments` that it's comprised of, and join them together.

```go
var code bytes.Buffer
lines := cb.Lines()
for i := 0; i < lines.Len(); i++ {
	line := lines.At(i)
	code.Write(line.Value(src))
}

fmt.Println("Code:\n", code.String())
```

**Comment by shazib-summar** [comment]
@abhinav hey thanks for the explanation super helpful. can you please elaborate how some nodes have `Lines` func defined on them but some dont? How do I tell which node kinds have Lines and which ones dont?

Also whats the difference between Text Block, and Paragraph? Similary Text vs String? I went through the docs but didnt find anything helpful there.

Thanks in advance 🙏🏾

**Reply by abhinav** [reply]
> How do I tell which node kinds have Lines and which ones dont

Only nodes with [Type()](https://pkg.go.dev/github.com/yuin/goldmark/ast#Node.Type) == [TypeBlock](https://pkg.go.dev/github.com/yuin/goldmark/ast#TypeBlock) will have `Lines` associated with them.

> difference between Text Block, and Paragraph

Text is an inline node, Paragraph is a block node.
A paragraph will typically have any number of inline nodes in it, including Text, Link, RawHTML, Emphasis, CodeSpan, etc.

> Text vs String

The difference between those two is the source of information:
Text refers to text in the original source that was provided to parser.Parse.
String owns the text in its Value field. This is useful if you're manipulating the AST and adding text to it.
You can use the [Node.Text](https://pkg.go.dev/github.com/yuin/goldmark/ast#Node.Text) method to get what's inside the Text or String in a generic way.
You'll notice the main difference between the two types in their implementations of that method:
[Text.Text()](https://github.com/yuin/goldmark/blob/v1.7.1/ast/inline.go#L146) returns a span of text from the provided `[]byte`.
[String.Text()](https://github.com/yuin/goldmark/blob/v1.7.1/ast/inline.go#L261) returns the value it has on its own Value field.
